core/src/test/java/io/forestframework/core/http/routing/HandlerIntegrationTest.java

```diff
@@ -0,0 +1,382 @@
+        assertThat(result, containsString(Message.ERROR_IN_CUSTOM_ERROR_HANDLER.name()));
+
+        assertThat(result, not(containsString("Resource not found")));
+    }
+
+    @Test
+    void exceptionsInPostHandlersAreHandledByFallbackErrorHandlerNotCustomHandler() throws IOException {
+        String result = sendHttpRequest("GET", "/errorInPostHandler/random").getBody();
+
+        assertThat(getTraces(), not(hasItem(Message.CUSTOM_ERROR_HANDLER.name())));
+        Assertions.assertEquals(
+                Collections.synchronizedList(Arrays.asList(
+                        Message.HANDLER.name(),
+                        Message.ERROR_IN_POSTHANDLER.name()
+                )), getTraces());
+
+        assertThat(result, not(containsString("Resource not found")));
+    }
+
+    @Test
+    void uncaught404ErrorIsHandledByFallbackErrorHandler() throws IOException {
+        String result = sendHttpRequest("GET", "/uncaught404/inexistence").getBody();
+
+        System.out.println(result);
+
+        assertThat(getTraces(), not(hasItem("uncaught404Error")));
+        assertThat(getTraces(), not(hasItem(Message.CUSTOM_ERROR_HANDLER.name())));
+        Assertions.assertEquals(HttpStatusCode.NOT_FOUND.name(), result);
+
+        assertThat(result, not(containsString("Resource not found")));
+    }
+
+    @Test
+    void httpExceptionCanBeThrownFromPreHandlersToPreventPropagation() throws IOException {
+        String result = sendHttpRequest("GET", "/prehandlerError/random").getBody();
+
+        System.out.println(result);
+        System.out.println(getTraces());
+        Assertions.assertEquals(
+                Arrays.asList(
+                        Message.ERROR_IN_PREHANDLER.name(),
+                        "shouldBeHandledByCustomErrorHandler",
+                        "shouldContinueInPostHandler"
+                ), getTraces());
+        assertThat(result, containsString(Message.ERROR_IN_PREHANDLER.name()));
+        assertThat(result, containsString("should be handled by custom error handler"));
+
+        assertThat(result, not(containsString("Resource not found")));
+    }
+
+    @Test
+    void shouldThrowExceptionsWhenPreHandlersNotReturnBoolean() throws IOException {
+        String result = sendHttpRequest("GET", "/preHandlerNotReturnBoolean/random").getBody();
+
+        System.out.println(result);
+        Assertions.assertEquals(
+                Arrays.asList(
+                        "prehandlerReturnsString",
+                        Message.CUSTOM_ERROR_HANDLER.name(),
+                        "shouldContinueInPostHandler"
+                ), getTraces());
+
+        assertThat(getTraces(), not(hasItem(Message.HANDLER.name())));
+        assertThat(result, containsString("java.lang.ClassCastException"));
+
+        assertThat(result, not(containsString("Resource not found")));
+    }
```

Comment:
We should throw errors at startup time, not runtime. I.e. inspect `Routing.getHandlerMethod()` return type via reflection, if the return type is not `void`/`boolean`/`Boolean`/`Future<Void>`/`CompletableFuture<Void>`/`Future<Boolean>`/`CompletableFuture<Boolean>`, throw exception and abort.

Reply:
‘kay, will fix this.